src/Microsoft.Diagnostics.Monitoring/DiagnosticsMonitor.cs

```diff
@@ -19,359 +19,119 @@ namespace Microsoft.Diagnostics.Monitoring
 {
     public sealed class DiagnosticsMonitor : IAsyncDisposable
     {
-        private readonly IServiceProvider _services;
-        private readonly Microsoft.Extensions.Logging.ILogger<DiagnosticsMonitor> _logger;
         private readonly MonitoringSourceConfiguration _sourceConfig;
-        private readonly IEnumerable<IMetricsLogger> _metricLoggers;
+        private readonly CancellationTokenSource _disposeSource;
+        private readonly object _lock = new object();
+        private Task _currentTask;
+        private bool _disposed;
 
-        //These values don't change so we compute them only once
-        private readonly List<string> _dimValues;
-
-        public const string NamespaceName = "Namespace";
-        public const string NodeName = "Node";
-        private static readonly List<string> DimNames = new List<string>{ NamespaceName, NodeName};
-
-        private int _disposeState = 0;
-
-        public DiagnosticsMonitor(IServiceProvider services, MonitoringSourceConfiguration sourceConfig)
+        public DiagnosticsMonitor(MonitoringSourceConfiguration sourceConfig)
         {
-            _services = services;
             _sourceConfig = sourceConfig;
-            IOptions<ContextConfiguration> contextConfig = _services.GetService<IOptions<ContextConfiguration>>();
-            _dimValues = new List<string> { contextConfig.Value.Namespace, contextConfig.Value.Node };
-            _metricLoggers = _services.GetServices<IMetricsLogger>();
-            _logger = _services.GetService<ILogger<DiagnosticsMonitor>>();
+            _disposeSource = new CancellationTokenSource();
         }
 
-        public async Task ProcessEvents(int processId, CancellationToken cancellationToken)
+        public Task CurrentProcessingTask => _currentTask;
+
+        public Task<Stream> ProcessEvents(int processId, int duration, CancellationToken cancellationToken)
```

Comment:
I think you need to validate that the duration parameter is not an infinite timeout. Otherwise, a call to DisposeAsync would await forever unless the cancellationToken is set to cancel.

Reply:
Two things:
- DisposeAsync actually cancels first so it should work correctly.
- I do want to support infinite timeout here. I will make some tweaks to make sure that works properly.